Postponed kill() message with softReset

leptun opened the pull request:
This is a prototype implementation. Relevant comments copied from slack:

---

@leptun
I have a proposal that might fix a lot of our issues regarding kill():
Current situation. When kill is called, we turn off the interrupts, disable all heaters, show an LCD message and we halt. This has the problem that not all MCU peripherals are reset, pins stay in their previous state (see the BEEPER issue with the thermal model), the message can't be multi-page since the timebase is disabled at that point (no interrupts).
Proposal: We display the kill message just like with the crash dump message. So when kill gets called, we store the kill message pointer somewhere (might be uninitialized ram or eeprom), we trigger a softReset() so that a watchdog reset gets executed and all MCU peripherals/registers get reset. On startup we then check if the firmware reset due to kill() and we show the kill message early during startup, with the timebase and interrupts enabled. This solves both the beeper issue and the multi-page error message issue. I also think this approach is safer in general since it is guaranteed that watchdog brings the printer into a safe state, unlike whatever we do now in kill()

Now the question remains if this new approach would ask the user to reset the printer. I'd say no. In this case, the user should only need to press the knob to confirm the message and then the firmware should finish booting up.

And so that the octoprint users don't complain about the kill messages preventing bootup, the kill message pending flag (whatever that will be) should be cleared before showing the message so that a subsequent printer reset doesn't show the message again on startup.

---

The kill message pointer is stored in eeprom in this prototype.

leptun committed:
postponed kill() message with softReset

leptun committed:
Update eeprom documentation

wavexx commented on Firmware/Marlin_main.cpp:
```diff
@@ -1591,6 +1602,9 @@ void setup()
     // report crash failures
     fw_crash_init();
 
+    // report kill() events
```
I would move this much earlier during startup. Definitely before the wizard, so that a bug anywhere doesn't preclude the kill() message.

leptun replied:
I moved it earlier. This is as early as I'm confortable. It must be after st_init since that enables interrupts (needed by the full screen message).

leptun committed:
Move kill check earlier during startup

leptun committed:
Use a different location as a kill message pending flag

The old implementation would fail if the message was in progmem at address 0xffff or 0x0000 (both unlikely). It would also fail if the eeprom was initialized to some other random value, which could have been dangerous when displayed as a full screen message.

leptun commented:
Note: some of the kill messages need to be changed since a reset is no longer necessary. So instead of something like "reset to continue", it needs to be "click to continue". Or just remove that part entirely since it's a regular full screen message now and it's implied that you click to continue.

leptun committed:
Reorder eeprom reads

Saves 4B of flash

leptun committed:
Fix eeprom table addresses

3d-gussner reviewed (approved):
Tested on MK404 and I really like it.
